### PR TITLE
Bucket creation using tier

### LINKED ISSUE #1267: [Bucket Tier]Implementation of Selection Service for Bucket Tiering
**Issue/Feature Description:**
1. One the policy are selected by users, how backends are selected using this policy should be implemented

**Why this issue to fixed / feature is needed(give scenarios or use cases):**

**How to reproduce, in case of a bug:**

**Other Notes /  Environment Information: (Please give the env information, log link or any useful information for this issue)**

### PR DESCRIPTION
**Pull request by PravinRanjan10**
**What type of PR is this?**

1. User/Admin can create bucket using tiers. Now, Admin can hide the backends details from users.
2. This PR has only scope for Admin. i.e, Admin only can create bucket from tiers.

> /kind new feature
> /kind bug fix
> /kind cleanup
> /kind revert change
> /kind design
> /kind documentation
> /kind enhancement

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1267 

**Test Report Added?**:
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
> /kind TESTED
> /kind NOT-TESTED

**Test Report**:
1. Create Backend:
POST http://192.168.29.183:8089/v1/94b280022d0c4401bcf3b0ea85870519/backends

**Request Body:**
{
"Name": "admin-aws1",
"Type": "aws-s3",
"Region": "us-east-1",
"Access": "***************",
"Security": "**************",
"Endpoint": "s3.amazonaws.com"
}

2. Create Tiers:
POST http://192.168.29.183:8089/v1/94b280022d0c4401bcf3b0ea85870519/tiers

**Request Body:**
{
    "Name": "Gold", 
    "Backends": ["backendId"]
}

3. List Tiers:
GET http://192.168.29.183:8089/v1/94b280022d0c4401bcf3b0ea85870519/tiers

4. Create Bucket:

PUT http://192.168.29.183:8090/bucketName

**Request Body:**
   LocationConstraint>Gold</LocationConstraint
    
Additonal Header key: 
key = tier
value = Tier

Example:

headerbuck

Some Output Screenshot:
Create Tier:
CreateTier1

Create Bucket:
BucketCreate

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by rajat-soda** [comment]
Additonal Header key:
key = tier
value = Tier  ==> Should this be True/False ?

File: api/pkg/s3/bucketput.go [hunk 1]
```diff
@@ -75,7 +78,12 @@ func (s *APIService) BucketPut(request *restful.Request, response *restful.Respo
 			return
 		}
 
-		backendName := createBucketConf.LocationConstraint
+		var backendName string
+		if isTier == "True" {
```
Comment (rajat-soda) [inline api/pkg/s3/bucketput.go:82]: Pls make the check case-insensitive, to handle "true" / True / TRUE etc..

File: api/pkg/s3/bucketput.go [hunk 2]
```diff
@@ -75,7 +78,12 @@ func (s *APIService) BucketPut(request *restful.Request, response *restful.Respo
...
+			backendName = s.getBackendFromTier(ctx, createBucketConf.LocationConstraint)
+		} else {
+			backendName = createBucketConf.LocationConstraint
+		}
 		if backendName != "" {
```
Comment (rajat-soda) [inline api/pkg/s3/bucketput.go:87]: better check "" != backendName 
Is trim on the backendName needed ? if yes, please add trim
Please check in other places as well

File: api/pkg/s3/service.go [hunk 3]
```diff
@@ -160,3 +161,36 @@ func HandleS3Error(response *restful.Response, request *restful.Request, err err
...
+	var backendId, backendName string
+
+	tierResp, err := s.backendClient.ListTiers(ctx, &backend.ListTierRequest{})
+	if err != nil {
+		log.Error("list tier failed during getting backends from tier")
+		return ""
+	} else {
+		if len(tierResp.Tiers) > 0 {
+			for _, tier := range tierResp.Tiers {
+				if tier.Name == tierName {
```
Comment (rajat-soda) [inline api/pkg/s3/service.go:176]: Please check if trim is needed in string comparison .. in other places as well.

File: api/pkg/s3/service.go [hunk 4]
```diff
@@ -160,3 +161,36 @@ func HandleS3Error(response *restful.Response, request *restful.Request, err err
...
+					break
+				}
+			}
+		}
+	}
+
+	if backendId != "" {
+		backendRep, err := s.backendClient.GetBackend(ctx, &backend.GetBackendRequest{Id: backendId})
+		if err != nil {
+			log.Error("the selected backends from tier doesn't exists.")
```
Comment (kumarashit) [inline api/pkg/s3/service.go:187]: Will this error be reached if we have already validated backend while creating Tier? There cab be error due to some other issues

Comment (PravinRanjan10) [reply api/pkg/s3/service.go:187]: There could be chance, admin removed this backendsjust before reaching this line

File: api/pkg/s3/service.go [hunk 5]
```diff
@@ -160,3 +161,36 @@ func HandleS3Error(response *restful.Response, request *restful.Request, err err
 
 	return nil
 }
+
+func (s *APIService) getBackendFromTier(ctx context.Context, tierName string) string {
+	log.Info("The received tier name for getting backend name:", tierName)
+	var backendId, backendName string
+
+	tierResp, err := s.backendClient.ListTiers(ctx, &backend.ListTierRequest{})
```
[the remark this answers is not in the document]
Comment (PravinRanjan10) [reply api/pkg/s3/service.go:169]: Yes. possible. But currently here is only used. so i think it's ok for now

File: api/pkg/s3/bucketput.go [hunk 6]
```diff
@@ -31,6 +31,9 @@ import (
 )
 
 func (s *APIService) BucketPut(request *restful.Request, response *restful.Response) {
+	log.Info("Create bucket request received")
+	var isTier string
```
[the remark this answers is not in the document]
Comment (PravinRanjan10) [reply api/pkg/s3/bucketput.go:35]: it's a variable for checking, tier is enabled or not?. So i think it's ok.

File: api/pkg/s3/bucketput.go (hunk 1 above)
Comment (PravinRanjan10) [reply api/pkg/s3/bucketput.go]: Done. Now it's case-insensitive. Thanks

File: api/pkg/s3/service.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (PravinRanjan10) [reply api/pkg/s3/service.go:172]: I am making issues to track this.

**Review by kumarashit — APPROVED** [review]
LGTM

**Comment by rajat-soda** [comment]
LGTM

**Review by rajat-soda — APPROVED** [review]
LGTM

**Comment by anvithks** [comment]
@PravinRanjan10 Please link the appropriate issue for the PR.

**Review by anvithks — APPROVED** [review]
LGTM

**Pull request merged**